File: dsa/heaps/cpp/1_Heap.cpp
#include <iostream>
#include <vector>
#include <algorithm>

template <typename T>
class Heap
{
    std::vector<T> harr;
    int size;

    void max_heapify(int i)
    {
        auto left = 2 * i + 1;
        auto right = 2 * i + 2;

        auto greatest = i;
        if (left <= size && harr[greatest] < harr[left])
        {
            greatest = left;
        }
        if (right <= size && harr[greatest] < harr[right])
        {
            greatest = right;
        }

        if (greatest != i)
        {
            std::swap(harr[greatest], harr[i]);
            max_heapify(i);
        }
    }

    void min_heapify(int i)
    {

        auto left = 2 * i + 1;
        auto right = 2 * i + 2;

        auto smallest = i;
        if (left <= size && harr[smallest] > harr[left])
        {
            smallest = left;
        }
        if (right <= size && harr[smallest] > harr[right])
        {
            smallest = right;
        }

        if (smallest != i)
        {
            std::swap(harr[smallest], harr[i]);
            min_heapify(i);
        }
    }

    void percolateUp_max(int i)
    {
        int parent = (i - 1) / 2;

        if (i > 0 and harr[parent] < harr[i])
        {
            std::swap(harr[i], harr[parent]);
            percolateUp_max(parent);
        }
    }

public:
    Heap()
    {
        size = 0;
    }

    Heap(const std::vector<T> &v)
    {
        harr = v;
        buildHeap();
    }

    void buildHeap()
    {
    }

    void push(T elem)
    {

        harr.push_back(elem);
        percolateUp_max(size);

        ++size;
    }

    T pop()
    {
        if (empty())
            return -1;

        T maxElem = harr[0];
        harr[0] = harr[size - 1];
        harr.pop_back();
        max_heapify(0);

        size--;
        return maxElem;
    }

    T top()
    {
        return harr[0];
    }

    bool empty()
    {
        return size == 0;
    }

    int Size()
    {
        return size;
    }

    void print()
    {
        for (auto x : harr)
        {
            std::cout << x << " ";
        }
        std::cout << "\n";
    }
};

int main()
{
    Heap<int> h;
    h.push(1);
    h.push(4);
    h.push(6);
    h.push(0);
    h.push(9);
    h.push(0);

    std::cout << h.top() << std::endl;
    h.pop();

    h.print();
}

File: dsa/heaps/javascript/1_Heap.ts
class Heap<T>{
    private harr : T[];
    private count : number;
    
    private max_heapify(i:number){
        let left = 2*i +1;
        let right = 2*i + 2;

        let greatest = i;
        if(left<this.count && this.harr[greatest]<this.harr[left])
            greatest = left;
        if(right<this.count && this.harr[greatest]<this.harr[right])
            greatest = right;

        if(greatest!==i){
            [this.harr[i],this.harr[greatest]] = [this.harr[greatest],this.harr[i]];
            this.max_heapify(greatest);
        }
    }

    private perlocateUp_max(i:number){
        let parent = Math.floor((i-1)/2);

        if(i>0 && this.harr[parent]<this.harr[i] ){
            [this.harr[parent],this.harr[i]] = [this.harr[i],this.harr[parent]];
            this.perlocateUp_max(parent);
        }
    }

    private buildHeap(arr:T[]){

    }

    constructor(arr?:T[]){
        this.count = 0;
        this.harr = []
        if(arr!==undefined){
            // buildHeap
        }
    }

    public push(elem: T):void{
        //push the element to the end.
        this.harr.push(elem);
        //perlocate
        this.perlocateUp_max(this.count);
        this.count++;
    }

    public pop():T|null{
        if(this.count===0)
            return null;

        //store the first element.
        let greatestElem = this.harr[0];
        //assign the last element to the first element.
        this.harr[0] = this.harr[this.count-1];
        this.harr.pop();
        //heapify the root.
        this.max_heapify(0);

        this.count--;
        return greatestElem;

    }


    public top():T|null{
        if(this.empty())
            return null
        return this.harr[0];
    }

    public size():number{
        return this.count;
    }

    public empty():boolean{
        return this.count===0;
    }

    public print(){
        console.log(this.harr)
    }
}


let heap = new Heap<number>();
heap.push(1);
heap.push(4);
heap.push(6);
heap.push(0);
heap.push(9);
heap.push(0);
heap.pop();


heap.print();
